File: src/pages/api/parse_and_prepare_json.py
import re
import json
import base64

def parse_content(content):
    parsed_data = {}
    remote_features = {}

    pattern = r'\*\*(.+?)\*\*: (.*?)(?=\n\*\*|$)'
    matches = re.findall(pattern, content, re.DOTALL)

    for key, value in matches:
        slug_key = key.strip().lower().replace(" ", "_")
        clean_value = value.strip().replace('\n', ' ').replace('  ', ' ')

        if slug_key in [
            "wi-fi_quality", "outlet_access", "noise_level", "seating_comfort",
            "natural_light", "stay_duration_friendliness", "food_&_drink_options",
            "bathroom_access", "parking_availability"
        ]:
            remote_features[slug_key.replace("wi-fi", "wi_fi").replace("food_&_drink_options", "food_drink_options")] = clean_value
        elif slug_key == "tags":
            parsed_data["tags"] = [tag.strip() for tag in clean_value.split(",")]
        elif slug_key == "restaurant_score":
            try:
                parsed_data[slug_key] = float(re.findall(r"[\d.]+", clean_value)[0])
            except:
                parsed_data[slug_key] = 0
        else:
            parsed_data[slug_key] = clean_value

    # Rename restaurant_name → name
    if "restaurant_name" in parsed_data:
        parsed_data["name"] = parsed_data.pop("restaurant_name")

    # Build unique slug: name + ZIP
    address = parsed_data.get("address", "")
    zip_match = re.search(r"\b\d{5}(?:-\d{4})?\b", address)
    zip_or_city = zip_match.group(0) if zip_match else address.split(',')[1].strip().lower().replace(' ', '-') if ',' in address else 'unknown'

    name_slug = parsed_data["name"].lower().replace("'", "").replace("&", "and")
    name_slug = re.sub(r'[^a-z0-9]+', '-', name_slug).strip('-')
    parsed_data["slug"] = f"{name_slug}-{zip_or_city}"

    # Attach remote features
    parsed_data["remote_work_features"] = remote_features

    # Fallback logo
    if "logo_url" not in parsed_data:
        parsed_data["logo_url"] = ""

    return parsed_data

# 🔁 Step 1: Parse response from Gemini
raw = input_data['response']
data = parse_content(raw)

# 🔁 Step 2: Extract Place ID from any field
available_keys = list(input_data.keys())
place_id = input_data.get('place_id') or input_data.get('Place ID') or next(
    (input_data[k] for k in available_keys if k.lower().strip() == "place_id"),
    None
)

if not place_id:
    raise ValueError(f"🚨 Missing 'place_id' in input_data. Keys: {available_keys}")

# 🔁 Step 3: Finalize output
data["google_place_id"] = place_id
json_str = json.dumps(data, indent=2)
base64_str = base64.b64encode(json_str.encode("utf-8")).decode("utf-8")

return {
    "parsed_json": data,
    "base64_output": base64_str,
    "slug": data["slug"],
    "filename": f"{data['slug']}.json"
}

File: src/pages/api/parse-and-prepare-json.js
import base64 from "base-64";

/**
 * Parses content into structured JSON
 */
function parseContent(content) {
  const parsedData = {};
  const remoteFeatures = {};

  const pattern = /\*\*(.+?)\*\*: (.*?)(?=\n\*\*|$)/gs;
  const matches = [...content.matchAll(pattern)];

  for (const match of matches) {
    const key = match[1].trim().toLowerCase().replace(/ /g, "_");
    const value = match[2].trim().replace(/\n/g, " ").replace(/  +/g, " ");

    if (
      [
        "wi-fi_quality",
        "outlet_access",
        "noise_level",
        "seating_comfort",
        "natural_light",
        "stay_duration_friendliness",
        "food_&_drink_options",
        "bathroom_access",
        "parking_availability"
      ].includes(key)
    ) {
      const remapped = key
        .replace("wi-fi", "wi_fi")
        .replace("food_&_drink_options", "food_drink_options");
      remoteFeatures[remapped] = value;
    } else if (key === "tags") {
      parsedData["tags"] = value.split(",").map((tag) => tag.trim());
    } else if (key === "restaurant_score") {
      const scoreMatch = value.match(/[\d.]+/);
      parsedData[key] = scoreMatch ? parseFloat(scoreMatch[0]) : 0;
    } else {
      parsedData[key] = value;
    }
  }

  if ("restaurant_name" in parsedData) {
    parsedData["name"] = parsedData["restaurant_name"];
    delete parsedData["restaurant_name"];
  }

  // Slug generation
  const address = parsedData.address || "";
  const zipMatch = address.match(/\b\d{5}(?:-\d{4})?\b/);
  const zipOrCity = zipMatch
    ? zipMatch[0]
    : address.includes(",")
      ? address.split(",")[1].trim().toLowerCase().replace(/\s+/g, "-")
      : "unknown";

  let nameSlug = parsedData.name?.toLowerCase().replace(/['&]/g, (c) =>
    c === "&" ? "and" : ""
  );
  nameSlug = nameSlug?.replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");

  parsedData.slug = `${nameSlug}-${zipOrCity}`;
  parsedData.remote_work_features = remoteFeatures;

  if (!parsedData.logo_url) {
    parsedData.logo_url = "";
  }

  return parsedData;
}

/**
 * Main handler function — usable as API route or importable function
 */
export default async function handler(req, res) {
  try {
    const input = req.body;

    if (!input || !input.response) {
      return res.status(400).json({ error: "Missing 'response' in request body" });
    }

    const data = parseContent(input.response);

    // Extract Place ID
    const placeId =
      input.place_id ||
      input["Place ID"] ||
      Object.entries(input).find(
        ([k]) => k.toLowerCase().trim() === "place_id"
      )?.[1];

    if (!placeId) {
      return res.status(400).json({
        error: `Missing 'place_id' in input. Keys: ${Object.keys(input).join(", ")}`
      });
    }

    data.google_place_id = placeId;
    const jsonStr = JSON.stringify(data, null, 2);
    const base64Str = base64.encode(jsonStr);

    return res.status(200).json({
      parsed_json: data,
      base64_output: base64Str,
      slug: data.slug,
      filename: `${data.slug}.json`
    });
  } catch (err) {
    return res.status(500).json({ error: err.message });
  }
}
